components/browser/state/src/main/java/mozilla/components/browser/state/reducer/SearchReducer.kt

```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.reducer
+
+import mozilla.components.browser.state.action.SearchAction
+import mozilla.components.browser.state.search.SearchEngine
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SearchState
+
+internal object SearchReducer {
+    /**
+     * [SearchAction] Reducer function for modifying [SearchState].
+     */
+    fun reduce(state: BrowserState, action: SearchAction): BrowserState {
+        return when (action) {
+            is SearchAction.UpdateSearchEngineListAction -> {
+                state.updateSearchEnginesMap(action.searchEngineList)
+            }
+            is SearchAction.UpdateCustomSearchEngineAction -> {
+                state.updateSearchEngine(action.searchEngine)
+            }
+            is SearchAction.RemoveCustomSearchEngineAction -> {
+                state.removeSearchEngine(action.searchEngineId)
+            }
+            is SearchAction.SetDefaultSearchEngineAction -> {
+                state.setDefaultSearchEngineAction(action.searchEngineId)
+            }
+        }
+    }
+}
+
+private fun BrowserState.updateSearchEnginesMap(
+    searchEngineList: List<SearchEngine>
+): BrowserState {
+    return copy(search = search.copy(
+        searchEngines = searchEngineList.map { it.id to it }.toMap()
+    ))
+}
+
+private fun BrowserState.updateSearchEngine(
+    searchEngine: SearchEngine
+): BrowserState {
+    val mutableSearchEngineMap = search.searchEngines.toMutableMap()
```

Comment:
fly-by nit: :) I'd prefer if we didn't use mutable state in the reducers, even if it's in local scopes. The method would also become shorter:

```
private fun BrowserState.updateSearchEngine(
    searchEngine: SearchEngine
): BrowserState {
    return copy(search = search.copy(
        searchEngines = search.searchEngines + (searchEngine.id to searchEngine)
    ))
}
```

Same for `removeSearchEngine` below.

Reply:
Example where we use an immutable map: https://github.com/mozilla-mobile/android-components/blob/master/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/MediaReducer.kt#L69

Reply:
Awesome, thanks :).